optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/ConstraintSubTree.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import static org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNodeType.FROM;
+import static org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNodeType.GROUPBY_COLLECTING_ONLY;
+import static org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNodeType.GROUPBY_MAPPING_AND_COLLECTING;
+import static org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNodeType.GROUPBY_MAPPING_ONLY;
+import static org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNodeType.JOIN;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNode;
+import org.optaplanner.core.impl.score.stream.drools.common.rules.RuleAssembler;
+
+final class ConstraintSubTree {
+
+    private final boolean isJoin;
+    private final ConstraintSubTree leftSubTree;
+    private final ConstraintSubTree rightSubTree;
+    private final List<ConstraintGraphNode> nodes;
+
+    public ConstraintSubTree(List<ConstraintGraphNode> nodesWithoutJoin) {
+        this.isJoin = false;
+        this.leftSubTree = null;
+        this.rightSubTree = null;
+        this.nodes = Collections.unmodifiableList(nodesWithoutJoin);
+        if (nodes.isEmpty()) {
+            throw new IllegalStateException("Node list may not be empty.");
+        }
+        if (nodes.get(0).getType() != FROM) {
+            throw new IllegalStateException("First node is not a From (" + nodes.get(0) + ").");
+        }
+    }
+
+    public ConstraintSubTree(ConstraintSubTree leftSubTree, ConstraintSubTree rightSubTree,
+            List<ConstraintGraphNode> joinAndOtherNodes) {
+        this.isJoin = true;
+        this.leftSubTree = Objects.requireNonNull(leftSubTree);
+        this.rightSubTree = Objects.requireNonNull(rightSubTree);
+        this.nodes = Collections.unmodifiableList(joinAndOtherNodes);
+        if (nodes.isEmpty()) {
+            throw new IllegalStateException("Node list may not be empty.");
```

Comment:
Can the user trigger this error?

If not, prefix the error message with "Impossible state", so for example "Impossible state: the nodeList must not be empty." It signals that we're 100% sure that this is an OptaPlanner bug.

Reply:
Yeah, there will be more places like that in this code. I'll go over it again.